### PR TITLE
[packchk] Add xml validation for input PDSCs (#506)

### PR DESCRIPTION
**Pull request by yvesll**
Signed-off-by: yvesll <toyears@hotmail.com>

**NOTE:
This PR did not include the mandatory option which moves PACK.xsd to <cmsis_toolbox>/etc, I do not know the deployment flow.**

### CODE REVIEW

**Comment by JonatanAntoni** [comment]
Thanks @yvesll for your contribution. Unfortunately, I see some challenged with this enhancement. 

First of all, I'd like to understand why you think the schema check should be integrated into PackChk, natively.

I added an explicit schema check into the gen-pack script library, see https://github.com/Open-CMSIS-Pack/gen-pack/blob/27feb09121b49fdc7bf95e6927df9a50454cdffb/gen-pack#L80-L121. This uses xmllint from libxml2 which is available under MIT license for all supported platforms.

If you still think the schema check should be integral part of packchk, we'd need to sort out open questions. The PR currently fails the checks. This could be due to linking Xerces library dynamically but not shipping the required library alongside the binary. So far, we link everything statically and accept larger binaries to prevent shared lib issues across different platforms and versions.

Cheers,
Jonatan

**Comment by yvesll** [comment]
HI, @JonatanAntoni, thanks for the comments.
For the failed checks, I am still exploring an elegant way to dynamically link the Xerces-c library and make sure it can work for all supported arches. Building the static lib may evidently increase the binary size but I appreciate it if we accept it.

For the motivation of this PR, you can see discussions in #506.

**Comment by JonatanAntoni** [comment]
@yvesll, thanks for clarifying this. I think we should accept the static linkage overhead as we want our CMSIS-Toolbox to be independent from a specific Linux distro. Perhaps changing the build to use static link would already solve the remaining issue.

File: tools/packchk/include/CreateModel.h [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include "Validate.h"
 #include "RteModelReader.h"
 #include "PackChk.h"
+#include "XmlValidator.h"
```
Comment (edriouk) [inline tools/packchk/include/CreateModel.h:13]: XmlValidator includes Xerces includes, they should not poison the code.

File: tools/packchk/include/CreateModel.h [hunk 2]
```diff
@@ -26,11 +27,15 @@ class CreateModel
   bool CheckForOtherPdscFiles(const std::string& pdscFullPath);
   bool AddPdsc(const std::string& pdscFile, bool bSkipCheckForOtherPdsc = false);
   bool AddRefPdsc(const std::set<std::string>& pdscRefFiles);
+  bool SetPackXsd(const std::string& packXsdFile);
   bool ReadAllPdsc();
   bool PrintPdscFiles(std::list<std::string>& pdscFiles);
 
 private:
   RteModelReader m_reader;
+  XmlValidator m_xmlValidator;
```
Comment (edriouk) [inline tools/packchk/include/CreateModel.h:36]: Making this a dynamically allocated pointer would solve two problems:
1) no need to include XmlValidator.h in the header, forward declaration is sufficient.
2) the object can be created only if needed (validation is enabled)

File: libs/xmlschemachecker/test/data/PACK.xsd [hunk 3]
```diff
@@ -0,0 +1,1980 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edriouk) [inline libs/xmlschemachecker/test/data/PACK.xsd:1]: Could we pull this file from its original location?
https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/tree/main/schema

File: libs/xmlschemachecker/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+list(TRANSFORM HEADER_FILES PREPEND include/)
+
+add_library(XmlValidator STATIC ${SOURCE_FILES} ${HEADER_FILES})
+
+set_property(TARGET XmlValidator PROPERTY
+        MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+
+target_include_directories(XmlValidator
+        PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include/
+        SYSTEM ${CMAKE_BINARY_DIR}/external/xerces-c
```
Comment (edriouk) [inline libs/xmlschemachecker/CMakeLists.txt:18]: all xerces includes should be made private and not exposed to outside the library.

**Review by edriouk — CHANGES_REQUESTED** [review]

**Comment by yvesll** [comment]
Can someone help me rerun the "CodeQL" check? I do not have access and it seems a network issue.

**Comment by yvesll** [comment]
@thorstendb-ARM, sorry, I click the re-request button but not sure why it removes you from the list, can you join the review again?

**Review by edriouk — APPROVED** [review]
Looks good.
@JonatanAntoni, could we disable static analysis for all external modules?

**Comment by edriouk** [comment]
Code QL issues are ignored because they mostly come from an external  module Xerces

**Comment by thorstendb-ARM** [comment]
@yvesll , why did you add this? I am trying to find out which is better, stopping after the first found error vs. reporting all errors including subsequent ones.
`m_domParser->setValidationConstraintFatal(true);`

**Comment by yvesll** [comment]
@thorstendb-ARM, this field was set to test **setExitOnFirstFatalError** in the development stage and I forgot to remove it. I think both of the two ways make sense in different use cases, so an extra option is better.

**Pull request merged**